cdc/generation.cc

```diff
@@ -834,16 +834,10 @@ future<> generation_service::check_and_repair_cdc_streams() {
                 latest, db_clock::now());
             should_regenerate = true;
         } else {
-            std::unordered_set<dht::token> gen_ends;
-            for (const auto& entry : gen->entries()) {
-                gen_ends.insert(entry.token_range_end);
-            }
-            for (const auto& metadata_token : tmptr->sorted_tokens()) {
-                if (!gen_ends.contains(metadata_token)) {
-                    cdc_log.warn("CDC generation {} missing token {}. Regenerating.", latest, metadata_token);
-                    should_regenerate = true;
-                    break;
-                }
```

Comment:
Why is this no longer needed? Don't we need to still check that each vnode + shard pair has exactly one stream?

Reply:
This is still needed.

First of all, let's focus on vnodes/tokens first.
The previous code would check that the tokens of the generation contained all tokens of the token ring.
We want this to be true always.
What we want to improve is that there are no additional tokens that are unnecessary, so we want to check whether the tokens of the generation are equal to the tokens of the token ring.
To check this we can combine two checks:
1. tokens of generation contains all tokens of token ring
2. size(tokens of token ring) == size(tokens of generation)

these two give us
tokens of generation == tokens of token ring

This PR added 2 but removed 1, which is wrong.

I think we should ignore shards for now and handle them in a follow-up, if ever, because they are not so easy to handle.
> Don't we need to still check that each vnode + shard pair has exactly one stream?

that may be impossible if there is a vnode and a shard which have no common tokens.

Reply:
> This is still needed.
> 
> First of all, let's focus on vnodes/tokens first. The previous code would check that the tokens of the generation contained all tokens of the token ring. We want this to be true always. What we want to improve is that there are no additional tokens that are unnecessary, so we want to check whether the tokens of the generation are equal to the tokens of the token ring. To check this we can combine two checks:
> 
> 1. tokens of generation contains all tokens of token ring
> 2. size(tokens of token ring) == size(tokens of generation)
> 
> these two give us tokens of generation == tokens of token ring
> 
> This PR added 2 but removed 1, which is wrong.

Makes sense.

> 
> I think we should ignore shards for now and handle them in a follow-up, if ever, because they are not so easy to handle.

Let's leave it for later (never).

> 
> > Don't we need to still check that each vnode + shard pair has exactly one stream?
> 
> that may be impossible if there is a vnode and a shard which have no common tokens.

Fair enough.

Reply:
OK, I restored the original criterion and added a strict size comparison.